db.go

```diff
@@ -352,19 +357,118 @@ type DB struct {
 	mu                 sync.RWMutex
 	models             map[reflect.Type]*Model
 	mappings           map[reflect.Type]fieldMap
+
+	// Function to add a deadline to a query.
+	deadlineQueryRewriter func(db *DB, query interface{}, millis int64) (queryWithDeadline interface{}, err error)
+}
+
+type DBOption func(*DB) error
+
+// When passed as an option to NewDB(), disables query deadlines.
+func QueryDeadlineNone(db *DB) error {
+	db.deadlineQueryRewriter = noopDeadlineQueryRewriter
+	return nil
+}
+
+// When passed as an option to NewDB(), enables query deadlines on Percona Server 5.6 or later.
+// Query deadlines only affect SELECT queries run via the Executor.Query function.
+// It us up to the underlying database engine to enforce the deadline.
+func QueryDeadlinePercona56(db *DB) error {
+	db.deadlineQueryRewriter = perconaDeadlineQueryRewriter
+	return nil
+}
+
+// When passed as an option to NewDB(), enables query deadlines on MySql 5.7.4 or later.
+// Query deadlines only affect SELECT queries run via the Executor.Query function.
+// It us up to the underlying database engine to enforce the deadline.
+func QueryDeadlineMySQL57(db *DB) error {
+	db.deadlineQueryRewriter = mySql57DeadlineQueryRewriter
+	return nil
+}
+
+// When passed as an option to NewDB(), determines whether to allow string queries.
+func AllowStringQueries(allow bool) DBOption {
+	return func(db *DB) error {
+		db.AllowStringQueries = allow
+		return nil
+	}
+}
+
+// When passed as an option to NewDB(), determines whether to ignore unmapped database columns.
+func IgnoreUnmappedCols(ignore bool) DBOption {
+	return func(db *DB) error {
+		db.IgnoreUnmappedCols = ignore
+		return nil
+	}
+}
+
+// When passed as an option to NewDB(), determines whether to ignore missing database columns.
+func IgnoreMissingCols(ignore bool) DBOption {
+	return func(db *DB) error {
+		db.IgnoreMissingCols = ignore
+		return nil
+	}
 }
 
-// NewDB creates a new DB from an sql.DB.
-func NewDB(db *sql.DB) *DB {
-	return &DB{
-		DB:                 db,
-		AllowStringQueries: true,
-		IgnoreUnmappedCols: true,
-		IgnoreMissingCols:  false,
-		Context:            context.Background(),
-		Logger:             nil,
-		models:             map[reflect.Type]*Model{},
-		mappings:           map[reflect.Type]fieldMap{},
+// When passed as an option to NewDB(), SetQueryLogger sets database logger to the given QueryLogger.
+func SetQueryLogger(logger QueryLogger) DBOption {
+	return func(db *DB) error {
+		db.Logger = logger
+		return nil
+	}
+}
+
+// NewDB creates a new DB from an sql.DB. Zero or more DBOptions may be passed in and will
+// be processed in order.
+func NewDB(db *sql.DB, options ...DBOption) (*DB, error) {
```

Comment:
Looks like we've never actually done a release. Do we specify a version number anywhere?